controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go

```diff
@@ -65,17 +72,81 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.KubernetesVersion, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.KubernetesVersion) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "kubernetesVersion"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.ControlPlaneEndpoint, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.ControlPlaneEndpoint) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "controlPlaneEndpoint"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "apiServer"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "controllerManager"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.Scheduler, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.Scheduler) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "scheduler"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "dns"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.CertificatesDir, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.CertificatesDir) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "certificatesDir"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "imageRepository"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.UseHyperKubeImage, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.UseHyperKubeImage) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "useHyperKubeImage"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.FeatureGates, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.FeatureGates) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "featureGates"))
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.ClusterName, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.ClusterName) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "clusterName"))
+	}
```

Comment:
This seems we might need to keep it up to date for the future? I wonder if we could take an approach saying like we don't allow any changes except for X? We could potentially use unstructured to compare changes between json objects rather than do field-by-field

Reply:
It will have to change when the types go from kubeadm v1beta1 to v1beta2 (which they should soon since we're getting deprecation warnings now 😄).

Yeah, the redundancy is super annoying. To expand on your suggestion, I'm assuming unstructured might have some kind of diff functionality we could use here and then we'd diff the two objects, see which fields changed and then deny or allow based on that?

I can see if unstructured would provide us with that information

Reply:
I suggested unstructured because we can easily remove "paths" from the objects that we'll allow mutations regardless, but that's not strictly necessary.

We might be able to leverage https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/patch.go#L87 to generate a patch and see if there are any diffs.

Reply:
reasonable. Appreciate the pushback against this wall of copy paste, i'll find a better solution! 😅

Reply:
btw I'm ok doing this for now, opening an issue and improve in v0.3.x